File: admin-frontend/src/services/customerService.ts
import { http } from './http'

export type CustomerServiceConversation = {
  id: number
  user_id: number
  user_nickname?: string | null
  target_type: 'merchant' | 'platform'
  merchant_id?: number | null
  merchant_name?: string | null
  product_id?: number | null
  product_name?: string | null
  order_id?: number | null
  order_no?: string | null
  status: string
  last_message_at?: string | null
  last_message?: string | null
  unread_count: number
  created_at: string
  updated_at: string
}

export type CustomerServiceMessage = {
  id: number
  conversation_id: number
  sender_type: 'user' | 'merchant' | 'platform'
  sender_id: number
  sender_name?: string | null
  content_type: 'text' | 'image'
  content: string
  image_urls: string[]
  is_read: boolean
  created_at: string
}

export type PageResult<T> = {
  list: T[]
  page: number
  page_size: number
  total: number
}

export const adminCustomerService = {
  listConversations(session: 'platform' | 'merchant', params?: { status?: string; page?: number; page_size?: number }) {
    return http.get<unknown, { data: PageResult<CustomerServiceConversation> }>('/admin/customer-service/conversations', {
      params,
      headers: { 'X-Admin-Session': session },
    })
  },

  listMessages(session: 'platform' | 'merchant', conversationId: number, params?: { page?: number; page_size?: number }) {
    return http.get<unknown, { data: PageResult<CustomerServiceMessage> }>(`/admin/customer-service/conversations/${conversationId}/messages`, {
      params,
      headers: { 'X-Admin-Session': session },
    })
  },

  sendMessage(session: 'platform' | 'merchant', conversationId: number, payload: { content: string; content_type?: 'text' | 'image'; image_urls?: string[] }) {
    return http.post<unknown, { data: CustomerServiceMessage }>(
      `/admin/customer-service/conversations/${conversationId}/messages`,
      {
        content_type: payload.content_type ?? 'text',
        content: payload.content,
        image_urls: payload.image_urls ?? [],
      },
      { headers: { 'X-Admin-Session': session } },
    )
  },

  closeConversation(session: 'platform' | 'merchant', conversationId: number) {
    return http.post<unknown, { data: CustomerServiceConversation }>(
      `/admin/customer-service/conversations/${conversationId}/close`,
      undefined,
      { headers: { 'X-Admin-Session': session } },
    )
  },
}

File: app/models/customer_service.py
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class CustomerServiceConversation(Base):
    __tablename__ = "customer_service_conversation"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("user.id"), index=True)
    target_type: Mapped[str] = mapped_column(String(20), default="merchant", index=True)
    merchant_id: Mapped[int | None] = mapped_column(ForeignKey("merchant.id"), nullable=True, index=True)
    product_id: Mapped[int | None] = mapped_column(ForeignKey("product.id"), nullable=True, index=True)
    order_id: Mapped[int | None] = mapped_column(ForeignKey("orders.id"), nullable=True, index=True)
    status: Mapped[str] = mapped_column(String(30), default="open", index=True)
    last_message_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
    )


class CustomerServiceMessage(Base):
    __tablename__ = "customer_service_message"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    conversation_id: Mapped[int] = mapped_column(ForeignKey("customer_service_conversation.id"), index=True)
    sender_type: Mapped[str] = mapped_column(String(20), index=True)
    sender_id: Mapped[int] = mapped_column(Integer, index=True)
    content_type: Mapped[str] = mapped_column(String(20), default="text")
    content: Mapped[str] = mapped_column(Text, default="")
    image_urls: Mapped[str] = mapped_column(Text, default="[]")
    is_read: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
